### PR TITLE
feat(gitlabci): used yaml parsing

### LINKED ISSUE #6659: Feature Request: Support extended Docker configuration options in gitlabci managers
**What would you like Renovate to be able to do?**
Renovate doesn't pick up our GitLab CI Services that use the extended syntax, for example:
```yaml
ui-test:
  stage: test
  services:
    - name: internal.docker.hub.company/selenium/standalone-firefox:3.141.59-xenon
      alias: standalone-firefox
    - name: internal.docker.hub.company/selenium/standalone-chrome:3.141.59-xenon
      alias: standalone-chrome
```

**Describe the solution you'd like**
Renovate should be able to parse this syntax. I debugged it and found out the whole parsing is done using regular expressions (instead of a YAML parser) and the regex simply doesn't support this format (and many of other syntactically valid variations).

**Describe alternatives you've considered**
Maybe we can switch to the simplified syntax and use the longer hostnames to access the services, avoiding extended options and not needing the additional support. I'm not sure if/how the 64/255 character-limit for hostname applies to docker services, though.

### LINKED ISSUE #9115: refactor extractPackageFile to use js-yaml
**What would you like Renovate to be able to do?**

From @viceice:

> @HonkingGoose Please open a new issue:

> Refactor `extractPackageFile` to use `js-yaml` and reuse already parsed object from `extractAllPackageFiles`.
> So we should extract a new function `parseDoc` to be reused by both extract functions.

**Did you already have any implementation ideas?**

~If we're going to be using `js-yaml` here, might be a good idea to migrate from `js-yaml` v3 to v4 first (see issue #8234)?~ That way you're doing the refactoring on top of up-to-date packages.

### LINKED ISSUE #14765: Gitlab: Image with quotes failed to parse
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.6.10

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.8

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

service images with quotes are not parsed correctly any more:

Minimal example:
```
services:
  - "mariadb:10.4.11"
```
Gets parsed as `mariadb:10.4.11"`.

### Relevant debug logs

<details><summary>Logs</summary>

```
Object {
              "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
              "currentDigest": undefined,
              "currentValue": "10.4.11\"",
              "datasource": "docker",
              "depName": "mariadb",
              "replaceString": "mariadb:10.4.11\"",
            }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### LINKED ISSUE #14843: Refactor gitlabci manager to use yaml parser
### Describe the proposed change(s).

Refactor the existing implementation - which uses custom line-based parsing with regex - to instead use YAML parsing.

### PR DESCRIPTION
**Pull request by hasanwhitesource**
## Changes

Got rid of line by line regex parsing and used `js-yaml` library to read the yaml content.

## Context

- Closes #6659
- Closes #9115
- Closes #14765
- Closes #14843

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 1]
```diff
@@ -57,7 +62,7 @@ describe('modules/manager/gitlabci/extract', () => {
       ]);
       expect(res).toMatchSnapshot();
       expect(res).toHaveLength(1);
-      expect(res[0].deps).toHaveLength(4);
+      expect(res[0].deps).toHaveLength(10);
```
Comment (rarkins) [inline lib/modules/manager/gitlabci/extract.spec.ts:65]: Were we missing a lot before? I.e. previous implementation was incomplete/buggy?

**Review by rarkins — COMMENTED** [review]
Is there any code which can be refactored and shared with the gitlabci-includes? Or do you suggest to keep it as copy/paste right now and then combine them into one manager after this PR?

File: lib/modules/manager/gitlabci/extract.spec.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/manager/gitlabci/extract.spec.ts:65]: no, there are 6 new services in fixture file.

File: lib/modules/manager/gitlabci/extract.spec.ts [hunk 2]
```diff
@@ -155,5 +160,74 @@ describe('modules/manager/gitlabci/extract', () => {
...
+          currentDigest: undefined,
+          currentValue: 'test2',
+          datasource: 'docker',
+          depName: 'image2',
+          depType: 'service-image',
+          replaceString: 'image2:test2',
+        },
+      ];
+      const services = ['image:test', 'image2:test2'];
+      expect(extractFromServices(undefined) === undefined).toBeTruthy();
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.spec.ts:211]:
```suggestion
      expect(extractFromServices(undefined)).toUndefined();
```
use simpler expectations.

File: lib/modules/manager/gitlabci/extract.ts [hunk 3]
```diff
@@ -2,96 +2,84 @@ import is from '@sindresorhus/is';
...
-  lines: string[],
-  lineNumber: number
-): { lineNumber: number; line: string } {
-  let ln = lineNumber;
-  while (
-    ln < lines.length - 1 &&
-    (commentsRe.test(lines[ln]) || aliasesRe.test(lines[ln]))
-  ) {
-    ln += 1;
+export function extractFromImage(image: Image): PackageDependency {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:11]:
```suggestion
export function extractFromImage(image: Image | undefined): PackageDependency {
```

File: lib/modules/manager/gitlabci/extract.ts [hunk 4]
```diff
@@ -2,96 +2,84 @@ import is from '@sindresorhus/is';
...
 }
 
-export function extractPackageFile(content: string): PackageFile | null {
+export function extractFromServices(services: Services): PackageDependency[] {
   const deps: PackageDependency[] = [];
+
+  if (is.undefined(services)) {
+    return undefined;
+  }
+  services.forEach((s) => {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:32]: use `for ... of` loop

File: lib/modules/manager/gitlabci/extract.ts [hunk 5]
```diff
@@ -2,96 +2,84 @@ import is from '@sindresorhus/is';
...
+    dep.depType = 'image';
+    return dep;
+  }
+  dep = getDep(image.name);
+  dep.depType = 'image-name';
+  return dep;
 }
 
-export function extractPackageFile(content: string): PackageFile | null {
+export function extractFromServices(services: Services): PackageDependency[] {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:26]:
```suggestion
export function extractFromServices(services: Services | undefined): PackageDependency[] {
```

File: lib/modules/manager/gitlabci/extract.ts [hunk 6]
```diff
@@ -2,96 +2,84 @@ import is from '@sindresorhus/is';
...
+    }
+    const dep = getDep(s.name);
+    dep.depType = 'service-image';
+    deps.push(dep);
+  });
+
+  return deps;
+}
+
+export function extractFromObject(prop: any, value: any): PackageDependency[] {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:47]: no any, use concrete types or unknown.

File: lib/modules/manager/gitlabci/extract.ts [hunk 7]
```diff
@@ -2,96 +2,84 @@ import is from '@sindresorhus/is';
...
-          if (serviceImageMatch) {
-            logger.trace('serviceImageMatch');
-            foundImage = true;
-            lineNumber = serviceImageLine.lineNumber;
-            const dep = getDep(serviceImageMatch.groups.depName);
-            dep.depType = 'service-image';
-            deps.push(dep);
-          }
-        } while (foundImage);
+    const doc: any = load(replaceReferenceTags(content), {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:64]: no any, define types.

File: lib/modules/manager/gitlabci/extract.ts [hunk 8]
```diff
@@ -2,96 +2,95 @@ import is from '@sindresorhus/is';
...
+export function extractFromServices(
+  services: Services | undefined
+): PackageDependency[] {
+  if (is.undefined(services)) {
+    return undefined;
+  }
+
   const deps: PackageDependency[] = [];
+  let dep: PackageDependency<Record<string, any>> = {};
+  for (const s of services) {
```
Comment (rarkins) [inline lib/modules/manager/gitlabci/extract.ts:35]:
```suggestion
  for (const service of services) {
```

Better to use longer variable names where possible

File: lib/modules/manager/gitlabci/extract.ts [hunk 9]
```diff
@@ -2,96 +2,95 @@ import is from '@sindresorhus/is';
...
+    if (is.string(s)) {
+      dep = getDep(s);
+      dep.depType = 'service-image';
+    } else {
```
Comment (rarkins) [inline lib/modules/manager/gitlabci/extract.ts:39]: Should you check if s.name is a string?

File: lib/modules/manager/gitlabci/extract.ts [hunk 10]
```diff
@@ -2,96 +2,95 @@ import is from '@sindresorhus/is';
...
+  if (is.undefined(value)) {
+    return undefined;
+  }
+  if (is.string(value) && property === 'image') {
+    return [extractFromImage(value as Image)];
+  }
+
+  let deps: PackageDependency[] = [];
+  const { image, services } = { ...(value as JobTemplate) };
+  if (!is.undefined(image)) {
```
Comment (rarkins) [inline lib/modules/manager/gitlabci/extract.ts:62]: Shouldn't this be a more complex check like is.plainObject()?

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/gitlabci/extract.ts [hunk 11]
```diff
@@ -2,96 +2,95 @@ import is from '@sindresorhus/is';
...
-          if (serviceImageMatch) {
-            logger.trace('serviceImageMatch');
-            foundImage = true;
-            lineNumber = serviceImageLine.lineNumber;
-            const dep = getDep(serviceImageMatch.groups.depName);
-            dep.depType = 'service-image';
-            deps.push(dep);
-          }
-        } while (foundImage);
+    const doc = load(replaceReferenceTags(content), {
```
[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/modules/manager/gitlabci/extract.ts:75]: unknown can't be set to a type , do you mean to cast it as a type?
I thought to create a type `type GitLabCiYml = Image | Services | JobTemplate;` to cast what comes from the load function. 
What you do think ?

[the remark this answers is not in the document]
Comment (hasanwhitesource) [reply lib/modules/manager/gitlabci/extract.ts:75]: Or to just check if what comes from the yml file is an object using `is.object`

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/gitlabci/extract.ts:75]: it should at least be an object and should be casted to an `Record<string, ....>`

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/gitlabci/extract.ts [hunk 12]
```diff
@@ -2,95 +2,90 @@ import is from '@sindresorhus/is';
...
+  if (is.undefined(services)) {
+    return [];
+  }
   const deps: PackageDependency[] = [];
+  for (const service of services) {
+    if (is.string(service)) {
+      const dep = getDep(service);
+      dep.depType = 'service-image';
+      deps.push(dep);
+    } else if (is.string(service.name)) {
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:40]:
```suggestion
    } else if (is.string(service?.name)) {
```

File: lib/modules/manager/gitlabci/extract.ts [hunk 13]
```diff
@@ -2,95 +2,90 @@ import is from '@sindresorhus/is';
...
+    return [];
+  }
+  let deps: PackageDependency[] = [];
+  if (is.object(job)) {
+    const { image, services } = { ...job };
+    if (is.object(image) || is.string(image)) {
+      deps.push(extractFromImage(image));
+    }
+    if (is.array(services)) {
+      deps = deps.concat(extractFromServices(services));
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:60]:
```suggestion
      deps.push(...extractFromServices(services));
```

File: lib/modules/manager/gitlabci/extract.ts [hunk 14]
```diff
@@ -2,95 +2,90 @@ import is from '@sindresorhus/is';
...
+    }) as Record<string, Image | Services | Job>;
+    if (is.object(doc)) {
+      for (const [property, value] of Object.entries(doc)) {
+        switch (property) {
+          case 'image':
+            deps.push(extractFromImage(value as Image));
+            break;
+
+          case 'services':
+            deps = deps.concat(extractFromServices(value as Services));
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:80]:
```suggestion
            deps.push(...extractFromServices(value as Services));
```

File: lib/modules/manager/gitlabci/extract.ts [hunk 15]
```diff
@@ -2,95 +2,90 @@ import is from '@sindresorhus/is';
...
+            break;
+
+          default:
+            deps = deps.concat(extractFromJob(value as Job));
```
Comment (viceice) [inline lib/modules/manager/gitlabci/extract.ts:84]:
```suggestion
            deps.push(...extractFromJob(value as Job));
```

**Review by viceice — COMMENTED** [review]
some small suggestions

**Review by viceice — APPROVED** [review]

**Pull request merged**